math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -100,27 +120,945 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
 
 namespace Internal {
 namespace VecOps {
+/// Return the next power of two (in 64-bits) that is strictly greater than A.
+/// Return zero on overflow.
+inline uint64_t NextPowerOf2(uint64_t A)
+{
+   A |= (A >> 1);
+   A |= (A >> 2);
+   A |= (A >> 4);
+   A |= (A >> 8);
+   A |= (A >> 16);
+   A |= (A >> 32);
+   return A + 1;
+}
+
+/// This is all the stuff common to all SmallVectors.
+///
+/// The template parameter specifies the type which should be used to hold the
+/// Size and Capacity of the SmallVector, so it can be adjusted.
+/// Using 32 bit size is desirable to shrink the size of the SmallVector.
+/// Using 64 bit size is desirable for cases like SmallVector<char>, where a
+/// 32 bit size would limit the vector to ~4GB. SmallVectors are used for
+/// buffering bitcode output - which can exceed 4GB.
+template <class Size_T>
+class SmallVectorBase {
+protected:
+   void *BeginX;
+   Size_T fSize = 0, Capacity;
+   bool Owns = true;
```

Comment:
Could this be `Capacity == -1`? I don't know what `Capacity == 12` and `fOwns == false` means - aren't these two mutually exclusive and this part of the same state description?

Reply:
we now use `fCapacity == 0` to indicate non-ownership. `clear()` calls `resetToSmall` in case of non-ownership, and `resetToSmall` now correctly sets `fCapacity` to the small vector capacity rather than to 0 as it happened in LLVM's SmallVector. The net effect is that `fCapacity == 0`  can now only be true if we are in memory adoption mode or if the small vector size is actually 0 (and it so happens that the behavior we want it these two cases is the same).